Replace IPath with java.nio.file.Path

tobous opened the pull request:
Replaces the usage of `IPath` with `java.nio.file.Path`. Adds `PathUtils` providing functionality previously provided by `IPath` but not provided by `java.nio.file.Path`. Removes `IPath` and `IPathFactory` as well as their implementations as they were no longer being used.

#### Reviewing This PR

I would suggest reviewing this PR commit by commit.


#### Commits

<details><summary><b>[INTERNAL] Prepare for replacing IPath with java.nio.file.Path</b></summary>
<br>

This commit is a preparation for the upcoming refactoring replacing the
usage of `IPath` with `java.nio.file.Path`.

Marks all methods in `IPath` and `IPathFactory` as deprecated. Adds notes
referencing the methods to use instead.

Adds `PathUtils`, offering utility methods to replace functionality
offered by `IPath` but not `java.nio.file.Path`. Also adds tests for the
added utility methods. The methods to make paths portable and convert
portable paths to `java.nio.file.Path` objects are explicitly excluded
from the tests as they rely on the local filesystem separator character.
This makes it very hard to write filesystem-independent tests that still
cover all cases.

Adds methods to convert between `java.nio.file.Path` and Eclipse `IPath`
objects to `ResourceConverter`. Marks the old conversion methods as
deprecated.

</details>

<details><summary><b>[FIX] #797 Replace the usage of IPath with java.nio.file.Path</b></summary>
<br>

Replaces the usage of `IPath` with `java.nio.file.Path` and `PathUtils`.

Resolves #797.

</details>

<details><summary><b>[NOP] Remove IPath and IPathFactory</b></summary>
<br>

Removes `IPath` as well as its implementations and their tests.

Removes `IPathFactory` as well as its implementations.

Removes `ResourceConverter` methods for `IPath` objects.

</details>

<details><summary><b>[REFACTOR] Clean up after IPath removal</b></summary>
<br>

Rename path conversion methods in `ResourceConverter` to match general
naming scheme. The used names were previously used by old
implementations.

Replaces full package references with package imports now that there is
only the Eclipse `IPath` interface.

</details>

<details><summary><b>[INTERNAL] Normalize path used for resource objects</b></summary>
<br>

Adjusts the existing Saros resource implementations to always normalize
the given paths. This ensures that Saros never operates on paths
pointing outside of the current workspace (without resorting to symlink
trickery). More specifically, it ensures that the relative paths
contained in the resource object can't be used to point to a parent
directory of the reference point once resolved.

</details>

tobous commented on core/src/saros/util/PathUtils.java:
```diff
@@ -0,0 +1,141 @@
+
+    Path normalized = path.normalize();
+
+    return normalized.getNameCount() == 1 && normalized.toString().equals("");
```
No, I just keep forgetting to use it. Will adjust.

tobous commented on core/src/saros/util/PathUtils.java:
```diff
@@ -0,0 +1,141 @@
+   * @return the remaining path
+   * @see Path#subpath(int, int)
+   */
+  public static Path removeFirstSegments(Path path, int count) {
```
I personally like the wording using "segment" better as it is easier to understand in my opinion. `java.nio.file.Path` uses the "namecount" wording, so I didn't really have a choice there. But calling this `dropFirstNameSegments(...)` seems very confusing to me.

tobous commented on core/src/saros/util/PathUtils.java:
```diff
@@ -0,0 +1,141 @@
+    String localPathString = path.toString();
+
+    if (File.separator.equals(PORTABLE_SEPARATOR)) {
+      if (localPathString.contains(WINDOWS_SEPARATOR)) {
```
Yeah, but I didn't want to overdo it with the special case handling. This can still be added later on if it seem important. 

Additionally, any of the forbidden characters should lead to an exception when trying to create a path containing them (at least I would hope so). This case, on the other hand, just leads to differing paths on the receiving side, which is much harder to notice/diagnose.

tobous commented on core/src/saros/util/PathUtils.java:
```diff
@@ -0,0 +1,141 @@
+
+    if (portablePathString.contains(WINDOWS_SEPARATOR)) {
+      log.warn(
+          "The given portable path contains occurrences of the Windows separator character '"
```
I will add the path to the log messages and adjust the content.

Regarding the issue of the current caller, I am not really sure what to do in that regard. I could throw an exception in this case or move this check to a separate method so that the caller would have to deal with it, but I am not whether that would be a better option.

tobous commented on server/test/junit/saros/server/filesystem/FileSystemTestUtils.java:
```diff
@@ -80,14 +65,13 @@ public static void createFile(ServerWorkspaceImpl workspace, String path) throws
 
-    IPath absolutePath = workspace.getLocation().append(path);
-    Path absoluteNioPath = nioPath(absolutePath);
+    Path absolutePath = workspace.getLocation().resolve(path);
```
The sever filesystem implementation needs improvement anyway (see #f980). Furthermore, I don't think that this patch changed anything in that regard. The server already used the java Path implementation to back the `IPath` implementation.

The best option for issues like these would probably be to include normalizing the given path as part of making it portable. This ensures that clients can't send such malicious paths to other participants.

srossbach replied:
I will test the Eclipse implementation tomorrow. Maybe it is possible to "leave" the reference point with manipulated data. If it is possible then this issue was already present with partial sharing but should not be reintroduced, to be more specific the normalization should also be done when generating file/folder handles or introduce a sanity check that the relative path is already normalized.

tobous commented:
Rebased onto current master without any interaction.

tobous reviewed (commented):
Amended the utility to always normalize paths before making them portable and after resolving paths from their portable representation.

Adds a normalization utility method. This method is now also used in the Resource implementations to normalize the given relative paths. The method normalizes the path and throws an exception if it still contains unresolved references to parent directories (i.e. starts with the segment `..`) after the normalization. This should prevent Saros resource objects being created for resources outside of the workspace scope.

tobous commented:
Rebased onto current master without any interaction.

tobous commented:
Added a note that all paths handled by resource objects must be normalized using `PathUtils.normalize(...)` to the `IResource` documentation.

Fixed an issue with `PathUtils.normalize(...)` where paths only containing an unresolved parent reference (just '`..`') were not correctly detected as illegal.

Added a test suite for `PathUtils`. The test suite explicitly does not cover `toPortableString(...)` and `fromPortableString(...)`. The methods rely on the local filesystem separator character, so I did not see a way of creating filesystem-independent tests that cover all possibilities.

srossbach commented on core/test/junit/saros/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,217 @@
+  @Test(expected = IllegalArgumentException.class)
+  public void testNormalizeDanglingParentReference() {
+    Path path = Paths.get("..");
+    System.out.println("Path: '" + PathUtils.normalize(path) + "'");
```
Assert.fail(...) ?

tobous replied:
The test fails anyways due to the missing exception. I just added to output to make debugging easier. These tests will have to be adjusted once we update to jUnit 4.13 anyway to use `assertThrows(...)` which allows specifying a message. See TODO at l.11.

srossbach commented on core/test/junit/saros/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,217 @@
+    Path path = Paths.get("foo", "..", "..", "bar");
+    System.out.println("Path: '" + PathUtils.normalize(path) + "'");
+  }
+}
```
Seems you are missing something like:   Paths.get("foo", "..", "..foobar")

tobous replied:
Done.
